contracts/exchange/contracts/src/MixinMatchOrders.sol

```diff
@@ -33,6 +33,101 @@ contract MixinMatchOrders is
+    /// @param leftSignatures Proof that left orders were created by the left makers.
+    /// @param rightSignatures Proof that right orders were created by the right makers.
+    /// @return batchMatchedFillResults Amounts filled and profit generated.
+    function batchMatchOrders(
+        LibOrder.Order[] memory leftOrders,
+        LibOrder.Order[] memory rightOrders,
+        bytes[] memory leftSignatures,
+        bytes[] memory rightSignatures
+    )
+        public
+        nonReentrant
+        returns (LibFillResults.BatchMatchedFillResults memory batchMatchedFillResults)
+    {
+        // Ensure that the left and right arrays are compatible and have nonzero lengths
+        require(leftOrders.length > 0, "Invalid number of left orders");
+        require(rightOrders.length > 0, "Invalid number of right orders");
+        require(leftOrders.length == leftSignatures.length, "Incompatible leftOrders and leftSignatures");
+        require(rightOrders.length == rightSignatures.length, "Incompatible rightOrders and rightSignatures");
+
+        // Without simulating all of the order matching, this program cannot know how many
+        // matches there will be. To ensure that batchMatchedFillResults has enough memory
+        // allocated for the left and the right side, we will allocate enough space for the
+        // maximum amount of matches (the maximum of the left and the right sides).
+        uint256 maxLength = _max256(leftOrders.length, rightOrders.length);
+        batchMatchedFillResults.left = new LibFillResults.FillResults[](maxLength);
+        batchMatchedFillResults.right = new LibFillResults.FillResults[](maxLength);
+
+        // Initialize initial variables
+        uint i;
+        uint256 leftIdx = 0;
+        uint256 rightIdx = 0;
+        LibOrder.Order memory leftOrder = leftOrders[0];
+        LibOrder.Order memory rightOrder = rightOrders[0];
+        bytes memory leftSignature = leftSignatures[0];
+        bytes memory rightSignature = rightSignatures[0];
+
+        // Loop infinitely (until broken inside of the loop), but keep a counter of how
+        // many orders have been matched.
+        for (i = 0;; i++) {
+            // Match the two orders that are pointed to by the left and right indices
+            LibFillResults.MatchedFillResults memory matchResults = _matchOrders(
+                leftOrder,
+                rightOrder,
+                leftSignature,
+                rightSignature
+            );
+
+            // Add the matchResults and the profit made during the match to the
+            // batchMatchedFillResults for this batch.
+            batchMatchedFillResults.left[i] = matchResults.left;
+            batchMatchedFillResults.right[i] = matchResults.right;
+            batchMatchedFillResults.profitInLeftMakerAsset +=
+                matchResults.left.makerFeePaid;
+            batchMatchedFillResults.profitInRightMakerAsset +=
+                matchResults.right.makerFeePaid;
```

Comment:
The fees aren't the profit. The profit is the spread between the two orders, the given to the taker (matcher). Currently we only support a spread in the left maker asset (that's what ZEIP-40 would address). So I think you want something like:
```solidity
batchMatchedFillResults.profitInLeftMakerAsset += matchResults.leftMakerAssetSpreadAmount;
batchMatchedFillResults.profitInRightMakerAsset += 0;
```

Should probably use `_safeAdd()` for these, as well.